pkg/scheduler/util/non_zero.go

```diff
@@ -34,45 +36,103 @@ const (
 	DefaultMilliCPURequest int64 = 100 // 0.1 core
 	// DefaultMemoryRequest defines default memory request size.
 	DefaultMemoryRequest int64 = 200 * 1024 * 1024 // 200 MB
+	// DefaultEphemeralStorage defines default empheral storage request size.
+	DefaultEphemeralStorage int64 = 0 // 0 Gi
 )
 
 // GetNonzeroRequests returns the default cpu and memory resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequests(requests *v1.ResourceList) (int64, int64) {
-	return GetNonzeroRequestForResource(v1.ResourceCPU, requests),
-		GetNonzeroRequestForResource(v1.ResourceMemory, requests)
+func GetNonzeroRequests(pod *v1.Pod) (int64, int64) {
+	return GetNonzeroRequestForResource(pod, v1.ResourceCPU),
+		GetNonzeroRequestForResource(pod, v1.ResourceMemory)
 }
 
 // GetNonzeroRequestForResource returns the default resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequestForResource(resource v1.ResourceName, requests *v1.ResourceList) int64 {
-	switch resource {
-	case v1.ResourceCPU:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceCPU]; !found {
-			return DefaultMilliCPURequest
+func GetNonzeroRequestForResource(pod *v1.Pod, resource v1.ResourceName) int64 {
+	if len(pod.Spec.Containers) <= 0 {
+		return 0
+	}
+
+	var podRequest int64
+	// The resource request may be assigned by the explicit 0. We need a state variable
+	// to indicate if the explicit resource request was found.
+	isExplictRequest := false
```

Comment:
I think we don't need to care whether the zero value is explicitly set or not. Any particular reason on this?

Reply:
https://github.com/kubernetes/kubernetes/blob/3538320d74f98fc0972206d20cd3d76186d88aa9/pkg/scheduler/framework/plugins/noderesources/least_allocated_test.go#L100

According to this test case, if we set an explicit zero resource request, we shouldn't give it a default value. That is reason

Reply:
sgtm

Reply:
To clarify: We should respect explicit 0.

However, in the existing behavior, we add non-zero value per container. We should preserve that. At least in this PR that we intent to patch.

Reply:
Not a fan of respecting explicit 0 - which doesn't really make sense to me. But it's ok to keep the behavior in this PR.